pkg/ds/daemon_set.go

```diff
@@ -205,9 +206,62 @@ func (ds *daemonSet) WatchDesires(
 		}
 	}()
 
+	// Send a node changed signal whenever the selected nodes in the store change
+	go func() {
+		changedCh := ds.watchNodeChanges(quitCh)
+		for {
+			select {
+			case <-quitCh:
+				return
+			case _, ok := <-changedCh:
+				if !ok {
+					// channel closed
+					return
+				}
+				nodesChangedCh <- struct{}{}
+			}
+		}
+	}()
+
 	return errCh
 }
 
+// Watch for changes to nodes and sends update and delete signals
+func (ds *daemonSet) watchNodeChanges(
+	quitCh <-chan struct{},
+) <-chan struct{} {
+	nodesChangedCh := make(chan struct{})
+
+	go func() {
+		var changes *labels.LabeledChanges
+
+		labeledChanges := ds.applicator.WatchMatchDiff(ds.NodeSelector, labels.NODE, quitCh)
+
+		for {
+			select {
+			case <-quitCh:
+				return
+			case val, ok := <-labeledChanges:
+				if !ok {
+					// Closed channel
+					return
+				}
+				changes = val
+				if changes == nil {
+					ds.logger.Error("Unexpected nil value received")
+					continue
+				}
+			}
+
+			if len(changes.Created) != 0 || len(changes.Updated) != 0 || len(changes.Deleted) != 0 {
+				nodesChangedCh <- struct{}{}
```

Comment:
Updated to ds smarter scheduling based on whether there was an create, update, or delete.